modules/module-postgres-storage/src/storage/PostgresBucketStorageFactory.ts

```diff
@@ -0,0 +1,526 @@
+  private sharedIterator = new sync.BroadcastIterable((signal) => this.watchActiveCheckpoint(signal));
+
+  private readonly storageCache = new LRUCache<number, storage.SyncRulesBucketStorage>({
+    max: 3,
+    fetchMethod: async (id) => {
+      const syncRulesRow = await this.db.sql`
+        SELECT
+          *
+        FROM
+          sync_rules
+        WHERE
+          id = ${{ value: id, type: 'int4' }}
+      `
+        .decoded(models.SyncRules)
+        .first();
+      if (syncRulesRow == null) {
+        // Deleted in the meantime?
+        return undefined;
+      }
+      const rules = new PostgresPersistedSyncRulesContent(this.db, syncRulesRow);
+      return this.getInstance(rules);
+    },
+    dispose: (storage) => {
+      storage[Symbol.dispose]();
+    }
+  });
+
+  constructor(protected options: PostgresBucketStorageOptions) {
+    super();
+    this.db = new lib_postgres.DatabaseClient({
+      config: options.config,
+      schema: STORAGE_SCHEMA_NAME,
+      notificationChannels: [NOTIFICATION_CHANNEL]
+    });
+    this.slot_name_prefix = options.slot_name_prefix;
+
+    this.db.registerListener({
+      connectionCreated: async (connection) => this.prepareStatements(connection)
+    });
+    this.notificationConnection = null;
+  }
+
+  async [Symbol.dispose]() {
+    await this.notificationConnection?.end();
+    await this.db[Symbol.asyncDispose]();
+  }
+
+  async prepareStatements(connection: pg_wire.PgConnection) {
+    // It should be possible to prepare statements for some common operations here.
+    // This has not been implemented yet.
+  }
+
+  getInstance(syncRules: storage.PersistedSyncRulesContent): storage.SyncRulesBucketStorage {
+    const storage = new PostgresSyncRulesStorage({
+      factory: this,
+      db: this.db,
+      sync_rules: syncRules,
+      batchLimits: this.options.config.batch_limits
+    });
+    this.iterateListeners((cb) => cb.syncStorageCreated?.(storage));
+    storage.registerListener({
+      batchStarted: (batch) => {
+        // This nested listener will be automatically disposed when the storage is disposed
+        batch.registerManagedListener(storage, {
+          replicationEvent: (payload) => this.iterateListeners((cb) => cb.replicationEvent?.(payload))
+        });
+      }
+    });
+    return storage;
+  }
+
+  async getStorageMetrics(): Promise<storage.StorageMetrics> {
+    const active_sync_rules = await this.getActiveSyncRules({ defaultSchema: 'public' });
+    if (active_sync_rules == null) {
+      return {
+        operations_size_bytes: 0,
+        parameters_size_bytes: 0,
+        replication_size_bytes: 0
+      };
+    }
+
+    const bucketData = await this.db.sql`
+      SELECT
+        --- This can differ from the octet_length
+        sum(pg_column_size(data)) AS operations_size_bytes
+      FROM
+        bucket_data
+      WHERE
+        group_id = ${{ type: 'int4', value: active_sync_rules.id }}
```

Comment:
These queries can be quite slow when the tables are large. Is there a faster way to get a size estimate for the entire table?

Reply:
I see I misread the MongoDB implementation. I assumed this was only for the active sync rules. If we want the entire table we can use `pg_total_relation_size`. Updated this to use `pg_total_relation_size` now.